src/Interpreters/ClusterDiscovery.cpp

```diff
@@ -339,24 +401,37 @@ bool ClusterDiscovery::updateCluster(ClusterInfo & cluster_info)
     }
 
     nodes_info = getNodes(zk, cluster_info.zk_root, node_uuids);
-    if (nodes_info.empty())
-    {
-        LOG_WARNING(log, "Can't get nodes info for '{}'", cluster_info.name);
-        return false;
-    }
 
     if (bool ok = on_exit(); !ok)
         return false;
 
     LOG_DEBUG(log, "Updating system.clusters record for '{}' with {} nodes", cluster_info.name, cluster_info.nodes_info.size());
 
-    auto cluster = makeCluster(cluster_info);
+    if (nodes_info.empty())
+    {
+        String name = cluster_info.name;
+        /// cluster_info removed inside removeCluster, can't use reference to name.
+        removeCluster(name);
+        return true;
+    }
 
+    auto cluster = makeCluster(cluster_info);
     std::lock_guard lock(mutex);
     cluster_impls[cluster_info.name] = cluster;
+
     return true;
 }
 
+void ClusterDiscovery::removeCluster(const String & name)
+{
+    {
+        std::lock_guard lock(mutex);
+        cluster_impls.erase(name);
+    }
+    clusters_to_update->remove(name);
+    LOG_DEBUG(log, "Dynamic cluster '{}' removed successfully", name);
```

Comment:
If we put `clusters_to_update->remove(name);` before `cluster_impls.erase(name);` would it solve issue with passing `cluster_info.name` to `removeCluster` ? If yes, are there other strong reason to order operations or not to like this?

Reply:
No, it can be in any order.
I did not remove record from `clusters_info` map, it's possible that at some moment we get cluster without nodes, when all nodes goes off, but we need to keep watch on this node in Keeper to restore cluster when nodes come back online. `name` is taken from `clusters_info`.